core/iwasm/aot/debug/jit_debug.c

```diff
@@ -0,0 +1,224 @@
+// and inline assembler statement inside.
+void __attribute__((noinline)) __jit_debug_register_code();
+
+void __attribute__((noinline)) __jit_debug_register_code()
+{
+    int x;
+    *(char *)&x = '\0';
+}
+
+// GDB will inspect contents of this descriptor.
+// Static initialization is necessary to prevent GDB from seeing
+// uninitialized descriptor.
+
+JITDescriptor __jit_debug_descriptor = { 1, JIT_NOACTION, NULL, NULL };
+#else
+extern void __jit_debug_register_code();
+extern JITDescriptor __jit_debug_descriptor;
+#endif
+
+// Call __jit_debug_register_code indirectly via global variable.
+// This gives the debugger an easy way to inject custom code to handle the events.
+void (*__jit_debug_register_code_ptr)() = __jit_debug_register_code;
+
+#ifdef __cplusplus
+}
+#endif
+
+typedef struct WASMJITDebugEngine {
+    korp_mutex jit_entry_mutex;
+    bh_list jit_entry_list;
+} WASMJITDebugEngine;
+
+typedef struct WASMJITEntryNode {
+    struct WASMJITEntryNode *next;
+    ;
+    JITCodeEntry *entry;
+} WASMJITEntryNode;
+
+static WASMJITDebugEngine *jit_debug_engine;
+
+static WASMJITDebugEngine *
+wasm_jit_debug_engin_create()
+{
+    WASMJITDebugEngine *engine;
+    if (!(engine = wasm_runtime_malloc(sizeof(WASMJITDebugEngine)))) {
+        LOG_ERROR("WASM Debug Engine error: failed to allocate memory");
+        return NULL;
+    }
+    memset(engine, 0, sizeof(WASMJITDebugEngine));
+    return engine;
+}
+
+void
+initial_jit_debug_engine()
+{
+    if (!jit_debug_engine)
+        jit_debug_engine = wasm_jit_debug_engin_create();
+    os_mutex_init(&jit_debug_engine->jit_entry_mutex);
+    bh_list_init(&jit_debug_engine->jit_entry_list);
+}
+
+static JITCodeEntry *
+CreateJITCodeEntryInternal(const uint8_t *symfile_addr, uint64_t symfile_size)
+{
+    JITCodeEntry *entry;
+
+    if (!jit_debug_engine)
+        return NULL;
+
+    os_mutex_lock(&jit_debug_engine->jit_entry_mutex);
+    if (!(entry = wasm_runtime_malloc(sizeof(JITCodeEntry)))) {
+        LOG_ERROR("WASM JIT Debug Engine error: failed to allocate memory");
+        os_mutex_unlock(&jit_debug_engine->jit_entry_mutex);
+        return NULL;
+    }
+    entry->symfile_addr_ = symfile_addr;
+    entry->symfile_size_ = symfile_size;
+    entry->prev_ = NULL;
+
+    entry->next_ = __jit_debug_descriptor.first_entry_;
+    if (entry->next_ != NULL) {
+        entry->next_->prev_ = entry;
+    }
+    __jit_debug_descriptor.first_entry_ = entry;
+    __jit_debug_descriptor.relevant_entry_ = entry;
+
+    __jit_debug_descriptor.action_flag_ = JIT_REGISTER_FN;
+
+    (*__jit_debug_register_code_ptr)();
+    os_mutex_unlock(&jit_debug_engine->jit_entry_mutex);
+    return entry;
+}
+
+static void
+DeleteJITCodeEntryInternal(JITCodeEntry *entry)
+{
+    if (!jit_debug_engine)
+        return;
+    os_mutex_lock(&jit_debug_engine->jit_entry_mutex);
+    if(entry->prev_ != NULL) { entry->prev_->next_ = entry->next_; }
```

Comment:
if (...) but not if(...), and had better put sentences inside { } into next lines.